### PR TITLE
Use nix-provided, dynmically linked rocksdb

### PR DESCRIPTION
**Pull request by elsirion**
By switching to a nix-provided rocksdb we can reduce compilation times, especially on first dev env setup since it can be fetched from cachix.

### CODE REVIEW

**Comment by justinmoon** [comment]
Android CI builds didn't finish. Re-ran them just in case it was flakiness. Perhaps we need to add them via flake.nix?

File: flake.nix [hunk 1]
```diff
@@ -273,6 +331,8 @@
           };
 
           LIBCLANG_PATH = "${pkgs.libclang.lib}/lib/";
+          ROCKSDB_LIB_DIR = "${rocksdb-7}/lib/";
+          LD_LIBRARY_PATH = "${rocksdb-7}/lib/";
```
Comment (elsirion) [inline flake.nix:335]: There's probably a better way to do this like `patchelf`ing the binary, but idk if that's practical here.

File: flake.nix [hunk 2]
```diff
@@ -622,6 +682,7 @@
         };
       in
       {
+        nixConfig.substituters = [ "https://fedimint.cachix.org/" ];
```
Comment (dpc) [inline flake.nix:685]: Putting in :tinfoilhat: . Not sure if we should do it here. It puts an extra trust in that host.

Bootstrap script already prompts users to do it: https://github.com/fedimint/fedimint/blob/0cf81c52305af181f787b07e704d2d5af35083eb/scripts/bootstrap.sh#L30 and then it's more of their own choice.

Comment (dpc) [reply flake.nix:685]: I kind of wonder if this even works. Do you have a link to where did you find it? I guess we'll know after it hits master.

File: flake.nix (hunk 1 above)
Comment (dpc) [reply flake.nix:335]: After a bit of investigation: This is caused by us using `mold` which is not wrapped to inject `rpath`.

This downgrades it to `lld` and fixes the issue:

```
diff --git a/.cargo/config.toml b/.cargo/config.toml
index b9fefd1f9..c4e86492f 100644
--- a/.cargo/config.toml
+++ b/.cargo/config.toml
@@ -1,6 +1,6 @@
 [target.x86_64-unknown-linux-gnu]
 linker = "clang"
-rustflags = ["-C", "link-arg=-fuse-ld=mold"]
+rustflags = ["-C", "link-arg=-fuse-ld=lld"]
 
 # mold is currently broken on MacOS
 # [target.aarch64-apple-darwin]
diff --git a/flake.nix b/flake.nix
index 272e5c94e..f09497643 100644
--- a/flake.nix
+++ b/flake.nix
@@ -308,6 +308,7 @@
             openssl
             pkg-config
             perl
+            pkgs.llvmPackages.bintools
             rocksdb-7
           ] ++ lib.optionals stdenv.isDarwin [
             libiconv
@@ -332,7 +333,6 @@
 
           LIBCLANG_PATH = "${pkgs.libclang.lib}/lib/";
           ROCKSDB_LIB_DIR = "${rocksdb-7}/lib/";
-          LD_LIBRARY_PATH = "${rocksdb-7}/lib/";
           CI = "true";
           HOME = "/tmp";
         };
```

Article about it: https://matklad.github.io/2022/03/14/rpath-or-why-lld-doesnt-work-on-nixos.html

**Comment by dpc** [comment]
Just thinking. Do we always prefer it dynamically linked? If we wanted we probably can keep it being built outside of the crate (better for build caching) yet still link statically. Should be possible.

File: flake.nix (hunk 2 above)
Comment (elsirion) [reply flake.nix]: Dropped the commit introducing it.

**Review by dpc — APPROVED** [review]

**Pull request merged**